### PR TITLE
JNG-4803 Deleting as a parent type

### PR DESCRIPTION
**Pull request by bencegelei**
<table>
<td>
  <a href="https://blackbelt.atlassian.net/browse/JNG-4803" title="JNG-4803" target="_blank">BugJNG-4803</a>  After deleting an entity instance as a parent type, the instance as a subtype is still present
  </td></table>
  <br />

### CODE REVIEW

File: judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java [hunk 1]
```diff
@@ -126,7 +123,7 @@ public void executeDeleteStatements(NamedParameterJdbcTemplate jdbcTemplate,
                                         " Params: " + ImmutableMap.copyOf(deleteStatementParameters.getValues()).toString());
 
                                 int count = jdbcTemplate.update(sql, deleteStatementParameters);
-                                checkState(count == 1, "There is illegal state, no records deleted");
+                                checkState(count == 0 || count == 1, "Maximum of 1 record should have been deleted. Actual: " + count);
```
Comment (robertcsakany) [inline judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java:127]: When zero record is presented, the message lies.

File: judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java [hunk 2]
```diff
@@ -101,17 +94,21 @@ public void executeDeleteStatements(NamedParameterJdbcTemplate jdbcTemplate,
...
 
-                    // Collecting all tables on the inheritance chain which required to insert.
-                    Stream.concat(
-                            ImmutableList.of(entity).stream(),
-                            entity.getEAllSuperTypes().stream()).filter(t -> isEntityType(t))
-                            .forEach(entityForCurrentStatement -> {
+                    // Collecting all tables on the inheritance chain to delete.
+                    Set<EClass> types = asmUtils.all(EClass.class).filter(c -> c.getEAllSuperTypes().contains(entity)).collect(Collectors.toSet());
+                    types.addAll(entity.getEAllSuperTypes());
+                    types.add(entity);
```
Comment (robertcsakany) [inline judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java]: Maybe better if AsmUtil have method where descandences can be query for an EClass. This recursive superclass collection for every class can be time consuming on every delete operation.

Comment (robertcsakany) [reply judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java]: And isEntityType check is missing.

**Review by robertcsakany — CHANGES_REQUESTED** [review]

File: judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java (hunk 1 above)
Comment (bencegelei) [reply judo-runtime-core-dao-rdbms/src/main/java/hu/blackbelt/judo/runtime/core/dao/rdbms/executors/DeleteStatementExecutor.java:127]: If no records were affected, there is no message.

**Review by noherczeg — APPROVED** [review]

**Review by robertcsakany — APPROVED** [review]

**Review by gaborflorian — APPROVED** [review]

**Pull request merged**